README


This archive contains the distribution Dist-Zilla-Plugin-Git,
version 2.029:

  update your git repository after release

This software is copyright (c) 2009 by Jerome Quelin.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.029.



Changes
Revision history for Dist::Zilla::Plugin::Git

2.029     2014-12-21 22:23:24+0100 Europe/Paris
 - don't "use Moose::Autobox" (Dist::Zilla doesn't use it anymore)
   (karen etheridge - github#22)

2.028     2014-11-30 05:48:59+0100 Europe/Paris
 - no functional changes
 - make tests more robust when GIT_* environment variables are set
   (christopher madsen)

2.027     2014-11-29 21:09:02+0100 Europe/Paris
 - unify string formatting code into Dist::Zilla::Role::Git::StringFormatter
   This allows you to use %c (the changelog entry) with Git::Tag.
   (christopher madsen, suggested by doug bell - rt#99990)

2.026     2014-11-22 17:35:45+0100 Europe/Paris
 - don't "use MooseX::AttributeShortcuts" (wasn't actually using its features)
   (christopher madsen, suggested by karen etheridge - rt#100498)

2.025     2014-10-11 19:05:56 Europe/Paris
 - add build_warnings option to Git::Check
   (christopher madsen, suggested by karen etheridge - rt#91057)

2.024     2014-09-13 21:21:42 Europe/Paris
 - add branch option to Git::Init (tuomas jormola - github#21)
 - Git::Init now logs when it makes the initial commit

2.023     2014-06-14 18:32:30 Europe/Paris
 - no functional changes
 - warn that Dist::Zilla::App::CommandHelper::ChainSmoking <= 1.04 is broken
   (karen etheridge - github#20)

2.022     2014-04-19 20:29:22 Europe/Paris
 - use MX:Path::Tiny Paths type to catch config issues at startup
   (kent fredric - github#18 & christopher madsen)
 - prune files using the prune_file method (karen etheridge - github#19)

2.021     2014-04-12 06:01:53 Europe/Paris
 - add support for MetaConfig (karen etheridge - github#17)
 - fix tests to work on Windows (christopher madsen)

2.020     2014-02-15 20:22:29 Europe/Paris
 - convert from Path::Class to Path::Tiny (kent fredric - github#16)
 - better documentation for allow_dirty_match (christopher madsen,
   requested by karen etheridge)

2.019     2013-11-23 20:49:27 Europe/Paris
 - fix #90537: wide character in print with ::Commit
   With Dist::Zilla 5, Git::Commit now writes the commit message in UTF-8
   (christopher madsen, reported by kent fredric)

2.018     2013-11-02 23:24:33 Europe/Paris
 TRIAL RELEASE
 - use the Config::Git plugin for shared configuration (brendan byrd)
   You don't have to use Dist::Zilla::Plugin::Config::Git unless you want to,
   but you do have to install it.

2.017     2013-10-26 19:57:59 Europe/Paris
 - remove use of deprecated Class::MOP function (karen etheridge)
 - make Git::CommitBuild compatible with Dist::Zilla 5
   (kent fredric, karen etheridge, & christopher madsen)

2.016     2013-10-19 18:27:32 Europe/Paris
 - fix exclude_filename to not match partial paths (karen etheridge)
   If you really intended to match a partial path, use exclude_match instead.
 - documentation typo fixed in Git::NextVersion (david steinbrunner)

2.015     2013-07-06 21:14:09 Europe/Paris
 - bail out of Build.PL unless git 1.5.4 or later is found (christopher madsen)
 - prevent warnings about chdir(undef) in tests (christopher madsen)

2.014     2013-06-22 17:50:26 Europe/Paris
 - fix rt#86251 - version-by-branch.t fails with Git 1.8.3
   The output of "git rev-list" changed slightly, and this broke
   Git::NextVersion when using version_by_branch=1.
   (christopher madsen, reported by zak elep)

2.013     2013-04-13 20:25:39 Europe/Paris
 - fix rt#84491 - Git::GatherDir does not handle symlinks
   Symlinks to directories used to make it crash with a cryptic error.
   It now skips them with a warning, and this is documented.
   (christopher madsen, reported by karen etheridge)
 - mention Git::GatherDir in Git (christopher madsen, reported by kevin dawson)

2.012     2013-03-27 00:41:08 Europe/Paris
 - add commit option to Git::Init (tatsuhiko miyagawa)

2.011     2013-03-23 16:19:52 Europe/Paris
 - fix compatibility with Git::Wrapper 0.028_96 (Don't pass Path::Class
   objects to Git::Wrapper) (christopher madsen)

2.010     2013-03-16 16:34:22 Europe/Paris
 - fix rt#83891 - Git::CommitBuild commits non-build files
   Git::CommitBuild now writes a fresh copy of the built tree to commit,
   instead of using one that may have been modified by other plugins
   (kent fredric & christopher madsen)

2.009     2013-02-08 06:03:33 Europe/Paris
 - add allow_dirty_match option to Git::Check and Git::Commit (graham knop)
 - change version-by-branch.t to use HEAD~1 instead of HEAD^
   Should fix test failure on Windows. (christopher madsen)
 - Test::Fatal 0.006 or better has been required since 2.007,
   but that wasn't explicit (christopher madsen, reported by jonathan leto)

2.008     2013-01-31 04:06:00 Europe/Paris
 - fix bug introduced in Git::Push 2.007 that broke remotes with
   additional parameters (christopher madsen)

2.007     2013-01-26 18:51:24 Europe/Paris
 - fix rt#82873 - Git::Push should check the remote exists
   New remotes_must_exist option defaults to true. (christopher madsen,
   suggested by olivier mengué)
 - switch from Test::Exception to Test::Fatal (christopher madsen)
 - load some dependencies only when needed (rt#82956) (christopher madsen,
   suggested by olivier mengué)

2.006     2012-12-29 23:17:39 Europe/Paris
 - document that the include_untracked option of Git::GatherDir
   requires Git 1.5.4, but has certain issues before Git 1.6.5.2.
   Adjust tests to skip when necessary. (christopher madsen)
 - set push.default matching in test config (christopher madsen)
   Otherwise Git for Windows 1.8.0 hangs in push-multi.t
   (thanks to christian walde for reporting the issue)

2.005     2012-12-08 22:03:15 Europe/Paris
 - avoid creating named type constraints (chris weyl & christopher madsen)

2.004     2012-12-01 19:58:59 Europe/Paris
 - improve Git::Commit's parsing of the Changes file (christopher madsen)
   Blank lines following the version line are no longer included in the commit
   message.  It can now find the current version even if followed by a -TRIAL

t/Util.pm
#
# This file is part of Dist-Zilla-Plugin-Git
#
# This software is copyright (c) 2009 by Jerome Quelin.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
#---------------------------------------------------------------------
package t::Util;
#
# Copyright 2012 Christopher J. Madsen
#
# Author: Christopher J. Madsen <perl@cjmweb.net>
# Created:  6 Oct 2012
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either the
# GNU General Public License or the Artistic License for more details.
#
# ABSTRACT: Utilities for testing Dist-Zilla-Plugin-Git
#---------------------------------------------------------------------

use 5.010;
use strict;
use warnings;

use Cwd qw(cwd);
use File::Copy::Recursive qw(dircopy);
use File::pushd qw(pushd tempd);
use Git::Wrapper ();
use Path::Tiny qw(path);
use Test::DZil qw(Builder);
use Test::More;
use version 0.80 ();

our ($base_dir, $base_dir_pushed, $dist_dir, $git_dir, $git, $zilla);

use Exporter ();
our @ISA    = qw(Exporter);
our @EXPORT = qw($base_dir $git_dir $git $zilla
                 append_and_add append_to_file init_test keep_tempdir
                 new_zilla_from_repo
                 skip_unless_git_version slurp_text_file
                 zilla_log_is);
our @EXPORT_OK = qw($dist_dir throws_ok zilla_version);

my $original_cwd;

BEGIN {
  # Change back to the original directory when shutting down,
  # to avoid problems with cleaning up tmpdirs.
  $original_cwd = path('.')->absolute;
  # we chdir around so make @INC absolute
  @INC = map {; ref($_) ? $_ : path($_)->absolute->stringify } @INC;
}

END { chdir $original_cwd if $original_cwd }

#=====================================================================
sub append_and_add
{
  my $fn = $_[0];

  &append_to_file;

  $git->add("$fn");
} # end append_and_add

#---------------------------------------------------------------------
sub append_to_file {
  my $file = shift;

  my $fh = $git_dir->child($file)->opena_utf8;
  print $fh @_;
  close $fh;
}

#---------------------------------------------------------------------
sub init_test
{
  my %opt = @_;

  $dist_dir = path('.')->absolute; # root of the distribution

  # Make a new directory so we don't affect the source repo:
  $base_dir_pushed = Path::Tiny->tempdir;
  $base_dir = $base_dir_pushed->absolute;

  # Mock HOME to keep user's global Git config from causing problems:
  mkdir($ENV{HOME} = $base_dir->child('home')->stringify)
      or die "Failed to create $ENV{HOME}: $!";

  delete $ENV{V}; # In case we're being released with a manual version
  delete $ENV{$_} for grep /^GIT_/i, keys %ENV;

  # Create the test repo:
  $git_dir = $base_dir->child('repo');
  $git_dir->mkpath;

  dircopy($dist_dir->child(corpus => $opt{corpus}), $git_dir)
      if defined $opt{corpus};

  if (my $files = $opt{add_files}) {
    while (my ($name, $content) = each %$files) {
      my $fn = $git_dir->child($name);
      $fn->parent->mkpath;
      $fn->spew_utf8( $content );
    }
  } # end if add_files

  my $pushd = pushd($git_dir);
  system "git init --quiet" and die "Can't initialize repo";

  $git = Git::Wrapper->new("$git_dir");

  $git->config( 'push.default' => 'matching' ); # compatibility with Git 1.8
  $git->config( 'user.name'  => 'dzp-git test' );
  $git->config( 'user.email' => 'dzp-git@test' );
} # end init_test

#---------------------------------------------------------------------
sub keep_tempdir
{
  $base_dir_pushed->preserve;
  print "Git files are in $git_dir\n";
} # end keep_tempdir

#---------------------------------------------------------------------
sub new_zilla_from_repo
{
  $zilla = Builder->from_config({dist_root => $git_dir}, @_);
} # end zilla_from_repo

#---------------------------------------------------------------------
our $git_version;

sub skip_unless_git_version
{
  my $need_version = shift;

  $git_version = version->parse(
    Git::Wrapper->new('.')->version =~ m[^( \d+ \. \d[.\d]+ )]x
  ) unless defined $git_version;

  if ( $git_version < version->parse($need_version) ) {
    my $why = "git $need_version or later required, you have $git_version";
    if (my $tests = shift) { skip $why, $tests     } # skip some
    else                   { plan skip_all => $why } # skip all
  } # end if we don't have the required version
} # end skip_unless_git_version

#---------------------------------------------------------------------
sub slurp_text_file
{
  my ($filename) = @_;

  return scalar do {
    local $/;
    # Don't use Path::Tiny's slurp_utf8 because it doesn't do
    # CRLF translation on Windows.
    if (open my $fh, '<:utf8', path( $zilla->tempdir )->child( $filename )) {
      <$fh>;
    } else {
      diag("Unable to open $filename: $!");
      undef;
    }
  };
} # end slurp_text_file

#---------------------------------------------------------------------
# DON'T USE THIS FUNCTION
#
# It's a limited clone of Test::Exception's throws_ok I wrote to make
# it easier to port the tests from Test::Exception to Test::Fatal.
# Use Test::Fatal directly instead.

sub throws_ok (&$;$)
{
  my ($coderef, $expected, $name) = @_;

  # The test program must load Test::Fatal
  my $exception = &Test::Fatal::exception($coderef);

  local $Test::Builder::Level = $Test::Builder::Level + 1;
  like( $exception, $expected, $name);
} # end throws_ok

#---------------------------------------------------------------------
sub zilla_log_is
{
  my ($matching, $expected, $name) = @_;

  $name //= "log messages for $matching";

  $matching = qr /^\Q[$matching]\E/ unless ref $matching;

  my $got = join("\n", grep { /$matching/ } @{ $zilla->log_messages });
  $got =~ s/\s*\z/\n/;

  local $Test::Builder::Level = $Test::Builder::Level + 1;
  is( $got, $expected, $name);

  $zilla->clear_log_events;
}

#---------------------------------------------------------------------
sub zilla_version
{
  my $pushd = pushd($zilla->root); # Must be in the correct directory
  my $version = $zilla->version;
  return $version;
} # end zilla_version

#=====================================================================
# Package Return Value:

1;

__END__


Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker;

WriteMakefile(
  DISTNAME  => 'Dist-Zilla-Plugin-Git',
  NAME      => 'Dist::Zilla::Plugin::Git',
  AUTHOR    => "Jerome\ Quelin",
  ABSTRACT  => "update\ your\ git\ repository\ after\ release",
  VERSION   => '1.093410',
  EXE_FILES => [ qw() ],
  (eval { ExtUtils::MakeMaker->VERSION(6.31) } ? (LICENSE => 'perl') : ()),
  PREREQ_PM    => {
    "MooseX::Has::Sugar" => '0',
    "MooseX::Types::Moose" => '0',
    "Dist::Zilla::Role::PluginBundle" => '0',
    "Git::Wrapper" => '0',
    "Dist::Zilla::Role::AfterRelease" => '0',
    "File::Find" => '0',
    "File::Path" => '2.07',
    "File::Spec::Functions" => '0',
    "Moose" => '0',
    "English" => '0',
    "Test::Exception" => '0',
    "Moose::Autobox" => '0',
    "Test::More" => '0',
    "File::Temp" => '0',
    "Dist::Zilla" => '1.093250',
    "Dist::Zilla::Role::BeforeRelease" => '0',
    "String::Formatter" => '0',
    "Cwd" => '0',
  },
  test => {TESTS => 't/*.t'}
);



Build.PL
#
# This file is part of Dist-Zilla-Plugin-Git
#
# This software is copyright (c) 2009 by Jerome Quelin.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#

use strict;
use warnings;

use Module::Build 0.3601;
use version 0.80 ();

my %module_build_args = (
  'dist_abstract' => 'update your git repository after release',
  'dist_author' => [
    'Jerome Quelin'
  ],
  'dist_name' => 'Dist-Zilla-Plugin-Git',
  'dist_version' => '2.029',
  'license' => 'perl',
  'module_name' => 'Dist::Zilla::Plugin::Git',
  'recursive_test_files' => 1,
  'build_requires' => {
    'Module::Build' => '0'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601',
    'version' => '0.80',
    'warnings' => '0'
  },
  'requires' => {
    'Cwd' => '0',
    'DateTime' => '0',
    'Dist::Zilla' => '4',
    'Dist::Zilla::Plugin::GatherDir' => '4.200016',
    'Dist::Zilla::Role::AfterBuild' => '0',
    'Dist::Zilla::Role::AfterMint' => '0',
    'Dist::Zilla::Role::AfterRelease' => '0',
    'Dist::Zilla::Role::BeforeRelease' => '0',
    'Dist::Zilla::Role::FilePruner' => '0',
    'Dist::Zilla::Role::GitConfig' => '0',
    'Dist::Zilla::Role::PluginBundle' => '0',
    'Dist::Zilla::Role::VersionProvider' => '0',
    'File::Spec' => '0',
    'File::Spec::Functions' => '0',
    'File::Temp' => '0',
    'File::chdir' => '0',
    'Git::Wrapper' => '0.021',
    'IPC::Open3' => '0',
    'IPC::System::Simple' => '0',
    'List::AllUtils' => '0',
    'List::Util' => '0',
    'Module::Runtime' => '0',
    'Moose' => '0',
    'Moose::Role' => '0',
    'Moose::Util::TypeConstraints' => '0',
    'MooseX::Has::Sugar' => '0',
    'MooseX::Types::Moose' => '0',
    'MooseX::Types::Path::Tiny' => '0.010',
    'Path::Tiny' => '0.048',
    'String::Formatter' => '0',
    'Try::Tiny' => '0',
    'Version::Next' => '0',
    'constant' => '0',
    'namespace::autoclean' => '0.09',
    'perl' => '5.010',
    'strict' => '0',
    'version' => '0.80',
    'warnings' => '0'
  },
  'test_requires' => {
    'CPAN::Meta::Check' => '0.007',
    'CPAN::Meta::Requirements' => '0',
    'Devel::SimpleTrace' => '0',
    'Dist::Zilla::File::InMemory' => '0',
    'Dist::Zilla::Plugin::Config::Git' => '0',
    'Dist::Zilla::Role::Releaser' => '0',
    'Dist::Zilla::Tester' => '0',
    'Encode' => '0',
    'Exporter' => '0',
    'File::Copy' => '0',
    'File::Copy::Recursive' => '0',
    'File::Path' => '2.07',
    'File::Spec' => '0',
    'File::Which' => '0',
    'File::pushd' => '0',
    'IO::Handle' => '0',
    'IPC::Open3' => '0',
    'List::MoreUtils' => '0',
    'Log::Dispatchouli' => '0',
    'Test::DZil' => '0',
    'Test::Fatal' => '0.006',
    'Test::More' => '0.88',
    'lib' => '0',
    'utf8' => '0'
  },
  get_options => { 'force-install' => undef },
);

unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

# Git 1.5.4 introduced the --exclude-standard option to git ls-files,
# which Git::Check needs to work properly.
my $need_version = version->parse('1.5.4');
eval {
  my $git_version = `git --version` or die "Unable to execute git\n";
  $git_version =~ /git \s+ version \s* ( \d+ (?: \.\d+ )+ )( [-.]rc\d+)?/x
      or die "Unable to determine git version\n";
  my $rc = $2 || '';

  $git_version = version->parse("$1");
  die "git $need_version or later required, you have $git_version$rc\n"
      if $git_version < $need_version or $git_version == $need_version and $rc;
  1;
} or do {
  print $@;

  if ($build->args('force-install')) {
    print "--force-install specified, attempting to install anyway.\n";
  } else {
    print <<"END MESSAGE";
\nDist::Zilla::Plugin::Git requires git $need_version or later in your PATH,
and it wasn't found.  If you want to install anyway, run

  $^X Build.PL --force-install
END MESSAGE

    exit 0;                  # Indicate missing non-Perl prerequisite
  } # end else not --force-install
}; # end do when Git prerequisite is not met

$build->create_build_script;
